sklearn/utils/validation.py

```diff
@@ -322,20 +336,34 @@ def check_array(array, accept_sparse=None, dtype="numeric", order=None,
     # store whether originally we wanted numeric dtype
     dtype_numeric = dtype == "numeric"
 
-    if sp.issparse(array):
-        if dtype_numeric:
+    dtype_orig = getattr(array, "dtype", None)
+    if not hasattr(dtype_orig, 'kind'):
+        # not a data type (e.g. a column named dtype in a pandas DataFrame)
+        dtype_orig = None
+
+    if dtype_numeric:
```

Comment:
I also thought about that option but I think it's fine to leave it this way for now.